exchangeBridge/exchange.go

```diff
@@ -0,0 +1,161 @@
+// via long-polling as opposed to a streaming model. This allows us to use TTRPC for the implementation, which currently
+// does not support streaming.
+type ExchangeBridge interface {
+	GetEvent(ctx context.Context) (*eventapi.Envelope, error)
+}
+
+type exchangeBridgeService struct {
+	eventChan <-chan *events.Envelope
+	errChan   <-chan error
+}
+
+// NewExchangeBridgeService returns a server-side implementation of the ExchangeBridge interface. Given an existing
+// exchange, it will subscribe to that exchange with no filters. Each GetEvent call pops and returns an event buffered
+// in the subscription's channel of published events.
+func NewExchangeBridgeService(ctx context.Context, eventSource events.Subscriber) ExchangeBridge {
+	eventChan, errChan := eventSource.Subscribe(ctx)
+	return &exchangeBridgeService{
+		eventChan: eventChan,
+		errChan:   errChan,
+	}
+}
+
+func RegisterExchangeBridgeService(srv *ttrpc.Server, svc ExchangeBridge) {
+	srv.Register(ExchangeBridgeServiceName, map[string]ttrpc.Method{
+		GetEventMethodName: func(ctx context.Context, unmarshal func(interface{}) error) (interface{}, error) {
+			return svc.GetEvent(ctx)
+		},
+	})
+}
+
+func (s *exchangeBridgeService) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	var (
+		envelope *eventapi.Envelope
+		err error
+	)
+
+	select {
+	case receivedEnvelope := <-s.eventChan:
+		envelope = &eventapi.Envelope{
+			Timestamp: receivedEnvelope.Timestamp,
+			Namespace: receivedEnvelope.Namespace,
+			Topic:     receivedEnvelope.Topic,
+			Event:     receivedEnvelope.Event,
+		}
+	case err = <-s.errChan:
+	case <-ctx.Done():
+		if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+			err = ctx.Err()
+		}
+	}
+
+	return envelope, err
+}
+
+type exchangeBridgeClient struct {
+	rpcClient *ttrpc.Client
+}
+
+func NewExchangeBridgeClient(rpcClient *ttrpc.Client) ExchangeBridge {
+	return &exchangeBridgeClient{
+		rpcClient: rpcClient,
+	}
+}
+
+func (c *exchangeBridgeClient) GetEvent(ctx context.Context) (*eventapi.Envelope, error) {
+	var req types.Empty
+	var resp eventapi.Envelope
+	if err := c.rpcClient.Call(ctx, ExchangeBridgeServiceName, GetEventMethodName, &req, &resp); err != nil {
+		return nil, err
+	}
+
+	return &resp, nil
+}
+
+// AttachExchangeBridge takes an existing ExchangeBridge and forwards each event retrieved over that bridge to the
+// provided forwarder. For example, if the ExchangeBridge is a client for a remote exchange and forwarder is a local
+// exchange, AttachExchangeBridge will result in each event published on the remote exchange to be forwarded to the
+// local exchange, essentially bridging the two together.
+func AttachExchangeBridge(ctx context.Context, exchangeBridge ExchangeBridge, forwarder events.Forwarder) error {
+	for {
+		select {
+		case <-ctx.Done():
+			if ctxErr := ctx.Err(); ctxErr != context.Canceled {
+				return ctxErr
+			} else {
+				return nil
+			}
+		default:
```

Comment:
Do we need empty default here? Without `default` "select" statement just blocks until ctx.Done can proceed

Reply:
We want the empty default so we don't get blocked on the call to ctx.Done(). The select here is just a check to see if the context was canceled already, at which point we can return from this function. If the context wasn't canceled, we want to continue on with the loop.

Reply:
Ah. I get it the idea. I thought this is a cycle :( I would rather suggest something like this instead:
```
if _, ok := <- ctx.Done(); ok {
			
}
```

Reply:
I completely forgot you could do that to implement a non-blocking read from a channel. Agree it's more readable, will update the PR

Reply:
Apparently that syntax actually is still blocking: https://play.golang.org/p/ccsYmAn4BTf

So I'll leave it as is but add a command above the select clarifying what it's doing